File: internal/http_client.go
package internal

import (
	"io"
	"log"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"strings"
	"time"

	"github.com/pkg/errors"
)

type HttpClient struct {
	client           *http.Client
	config           Config
	cookies          []*http.Cookie
	hasMoodleSession bool
}

func NewClient(config Config) (*HttpClient, error) {
	cookieJar, err := cookiejar.New(nil)
	if err != nil {
		return nil, errors.Wrap(err, "failed to initialise http client")
	}

	client := &http.Client{
		Jar:     cookieJar,
		Timeout: time.Minute,
	}

	return &HttpClient{client: client, config: config}, nil
}

func (h *HttpClient) newRequest(method string, url string, data io.Reader) (*http.Request, error) {
	request, err := http.NewRequest(method, url, data)
	if err != nil {
		return nil, errors.Wrap(err, "failed to create a new request")
	}

	request.Header.Set("Access-Control-Expose-Headers", "Set-Cookie")
	request.Header.Set("Sec-Fetch-Dest", "document")
	request.Header.Set("Sec-Fetch-Mode", "navigate")
	request.Header.Set("Sec-Fetch-Site", "none")
	request.Header.Set("Sec-GPC", "1")
	// request.Header.Set("Origin", h.config.SiakadUrl)
	// request.Header.Set("Referer", h.config.SiakadUrl)
	request.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36")

	for _, cookie := range h.cookies {
		// we don't want this cookie to override the actual `siakad` cookie
		isInvalidSiakadCookie := cookie.Name == "siakad" && cookie.Value == "a%3A0%3A%7B%7D"
		isDeletedPolinemaSSO := cookie.Name == "polinema_sso" && cookie.Value == "deleted"
		if isInvalidSiakadCookie || isDeletedPolinemaSSO {
			continue
		}
		request.AddCookie(cookie)
	}

	return request, nil
}

func (h *HttpClient) Get(url string, data io.Reader) (*http.Response, error) {
	request, err := h.newRequest(http.MethodGet, url, data)
	if err != nil {
		return nil, err
	}

	response, err := h.client.Do(request)
	if err != nil {
		return nil, errors.Wrap(err, "failed to perform a get request")
	}

	return response, err
}

func (h *HttpClient) postForm(url string, data url.Values) (*http.Response, error) {
	encodedData := data.Encode()
	request, err := h.newRequest(http.MethodPost, url, strings.NewReader(encodedData))
	if err != nil {
		return nil, err
	}

	request.Header.Set("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8")
	request.Header.Set("Accept", "application/json")
	request.Header.Set("Referer", h.config.SiakadUrl+"/login/index/err/6")
	request.Header.Set("X-Requested-With", "XMLHttpRequest")

	response, err := h.client.Do(request)
	if err != nil {
		return nil, errors.Wrap(err, "failed to perform a post form request")
	}

	return response, nil
}

func (h *HttpClient) login() error {
	credential := url.Values{
		"username": []string{h.config.NIM},
		"password": []string{h.config.Password},
	}
	response, err := h.postForm(h.config.SiakadUrl+"/login", credential)
	if err != nil {
		return errors.Wrap(err, "failed to login")
	}
	defer response.Body.Close()

	h.updateOrInsertCookie(response.Cookies())
	return nil
}

func (h *HttpClient) FetchSubjectsContent() (io.Reader, error) {
	response, err := h.Get(h.config.SlcUrl+"/spada", nil)
	if err != nil {
		return nil, err
	}
	h.updateOrInsertCookie(response.Cookies())
	return response.Body, nil
}

func (h *HttpClient) FetchLmsContent(courseUrl string) (io.Reader, error) {
	// the first firstResponse, which has a url format of `slcUrl/spada/?gotourl=xxx` is used to get the cookie needed for the lms page itself
	// the firstResponse is a <script>window.location="lmsUrl"</script>, which we do in the second request
	// not sure why they use client side redirect instead of responding with 302 status code
	firstResponse, err := h.Get(h.config.SlcUrl+"/spada/?gotourl="+url.QueryEscape(courseUrl), nil)
	if err != nil {
		return nil, err
	}
	defer firstResponse.Body.Close()
	h.updateOrInsertCookie(firstResponse.Cookies())

	// we need to get the MoodleSession first and then enter the url using that cookie
	// skip if we already have MoodleSession cookie
	if !h.hasMoodleSession {
		secondReq, _ := h.newRequest(http.MethodGet, courseUrl, nil)
		secondReq.Header.Set("Sec-Fetch-Site", "same-site")
		secondReq.Header.Set("Referer", h.config.SlcUrl)
		moodleSessionResp, err := h.client.Do(secondReq)
		if err != nil {
			return nil, err
		}
		moodleSessionResp.Body.Close()
		h.updateOrInsertCookie(moodleSessionResp.Cookies())

		// we can't simply do `h.hasMoodleSession = true` because we're not sure if it's succeeded or not
		for _, cookie := range h.cookies {
			if cookie.Name == "MoodleSession" {
				h.hasMoodleSession = true
				break
			}
		}
	}

	secondResponse, err := h.Get(courseUrl, nil)
	if err != nil {
		return nil, err
	}
	return secondResponse.Body, nil
}

func (h *HttpClient) CollectCookies() error {
	log.Println("collecting cookies...")
	if err := h.collectSiakadCookies(); err != nil {
		return err
	}

	log.Println("logging in...")
	if err := h.login(); err != nil {
		return err
	}

	log.Println("collecting homepage cookies...")
	if err := h.collectHomepageCookies(); err != nil {
		return err
	}

	log.Println("collecting slc cookies...")
	if err := h.collectSlcCookies(); err != nil {
		return err
	}

	return nil
}

func (h *HttpClient) collectSiakadCookies() error {
	response, err := h.Get(h.config.SiakadUrl, nil)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	h.updateOrInsertCookie(response.Cookies())
	return nil
}

func (h *HttpClient) collectHomepageCookies() error {
	response, err := h.Get(h.config.SiakadUrl+"/beranda", nil)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	h.updateOrInsertCookie(response.Cookies())
	return nil
}

func (h *HttpClient) collectSlcCookies() error {
	response, err := h.Get(h.config.SlcUrl, nil)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	h.updateOrInsertCookie(response.Cookies())
	return nil
}

func (h *HttpClient) ResetCookies() {
	h.cookies = []*http.Cookie{}
}

func (h *HttpClient) updateOrInsertCookie(newCookies []*http.Cookie) {
	for _, newCookie := range newCookies {
		oldIndex := -1
		for i, oldCookie := range h.cookies {
			if oldCookie.Name == newCookie.Name {
				oldIndex = i
			}
		}

		if oldIndex != -1 {
			h.cookies[oldIndex] = newCookie
		} else {
			h.cookies = append(h.cookies, newCookie)
		}
	}
}

File: worker-api/src/application/HttpClient.ts
import { FetchContext, type FetchOptions, Headers, ofetch } from "ofetch";
import type { ICookieJar } from "./interfaces/ICookieJar";
import type { ILogger } from "./interfaces/ILogger";
import type { Env } from "~/env";

export class HttpClient {
	#env: Env;
	#logger: ILogger;
	#cookieJar: ICookieJar;
	#hasMoodleSession = false;

	constructor(env: Env, logger: ILogger, cookieJar: ICookieJar) {
		this.#env = env;
		this.#logger = logger;
		this.#cookieJar = cookieJar;
	}

	private async collectSiakadCookies() {
		await this.fetch(this.#env.SIAKAD_URL, { method: "HEAD" });
	}

	private async collectHomepageCookies() {
		await this.fetch(`${this.#env.SIAKAD_URL}/beranda`, { method: "HEAD" });
	}

	private async collectSlcCookies() {
		// this needs to be done twice because for some reason
		// they decided to do this weird mechanism, i don't know why
		await this.fetch(this.#env.SLC_URL, { method: "HEAD" });
		await this.fetch(this.#env.SLC_URL, { method: "HEAD" });
	}

	private async login() {
		const formData = new FormData();
		formData.set("username", this.#env.NIM);
		formData.set("password", this.#env.PASSWORD);

		await this.postForm(`${this.#env.SIAKAD_URL}/login`, {
			body: formData,
		});
	}

	private async postForm(url: RequestInfo, options: FetchOptions = {}) {
		const headers = new Headers();
		headers.set("Accept", "application/json");
		headers.set("Referer", `${this.#env.SIAKAD_URL}/login/index/err/6`);
		headers.set("X-Requested-With", "XMLHttpRequest");
		return this.fetch(url, {
			...options,
			method: "POST",
			headers,
		});
	}

	private async fetch(
		url: RequestInfo,
		options: FetchOptions & {
			excludedCookies?: string[];
		} = {},
	) {
		// serialise the cookie
		const cookieString = this.#cookieJar
			.entries()
			.filter((cookie) => !options.excludedCookies?.includes(cookie.key))
			.map((cookie) => `${cookie.key}=${cookie.value}`)
			.join("; ");

		const headers = new Headers();
		headers.set("Cookie", cookieString);

		// merge headers if it's provided from the option
		if (options.headers !== undefined) {
			for (const [key, value] of (
				options.headers as Map<string, string>
			).entries()) {
				headers.set(key, value);
			}
		}

		// use chrome user agent by default
		headers.set(
			"User-Agent",
			"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
		);

		return ofetch(url, {
			...options,
			headers,
			onResponse: ({ response }) => {
				const cookies = response.headers.get("set-cookie");
				if (cookies === null) return;
				this.#cookieJar.store(response.headers.get("set-cookie") ?? "");
			},
		});
	}

	public async collectCookies() {
		try {
			this.#logger.info("collecting siakad cookies");
			await this.collectSiakadCookies();
			this.#logger.info("siakad cookies has been collected");
		} catch (err) {
			this.#logger.error("failed to collect siakad cookies");
			throw err;
		}

		try {
			this.#logger.info("attempting login");
			await this.login();
			this.#logger.info("successfully logged in");
		} catch (err) {
			this.#logger.error("failed to login");
			throw err;
		}

		try {
			this.#logger.info("collecting siakad homepage cookies");
			await this.collectHomepageCookies();
			this.#logger.info("siakad homepage cookie has been collected");
		} catch (err) {
			this.#logger.error("failed to collect siakad homepage cookie");
			throw err;
		}

		try {
			this.#logger.info("collecting slc cookies");
			await this.collectSlcCookies();
			this.#logger.info("slc cookie has been collected");
		} catch (err) {
			this.#logger.error("failed to collect slc cookie");
			throw err;
		}
	}

	public async fetchSubjectsContent() {
		const headers = new Headers();
		headers.set("Sec-Gpc", "1");
		headers.set("Sec-Fetch-Mode", "cors");
		headers.set("Sec-Fetch-Site", "same-origin");
		headers.set("Sec-Fetch-Dest", "empty");
		headers.set("Referer", `${this.#env.SLC_URL}/spada`);
		headers.set("Host", this.#env.SLC_URL);
		headers.set("Pragma", "no-cache");
		const response: string = await this.fetch(`${this.#env.SLC_URL}`, {
			parseResponse: (text) => text,
			headers,
		});
		return response;
	}

	public async fetchLmsContent(courseUrl: string) {
		// the first firstResponse, which has a url format of `slcUrl/spada/?gotourl=xxx` is used to get the cookie needed for the lms page itself
		// the firstResponse is a <script>window.location="<lmsUrl>"</script>, which we do in the second request
		// not sure why they use client side redirect instead of responding with 302 status code
		await this.fetch(
			`${this.#env.SLC_URL}/spada?gotourl=${encodeURIComponent(courseUrl)}`,
		);
		if (!this.#hasMoodleSession) {
			const headers = new Headers();
			headers.set("Sec-Fetch-Site", "same-site");
			headers.set("Referer", this.#env.SLC_URL);
			await this.fetch(courseUrl, { headers });

			// we can't simply do `h.hasMoodleSession = true` after fetching
			// because we can't be sure if it's succeeded or not
			if (this.#cookieJar.has("MoodleSession")) {
				this.#hasMoodleSession = true;
			}
		}

		return await this.fetch(courseUrl, { parseResponse: (text) => text });
	}

	public async fetchAssignmentContent(assignmentUrl: string) {
		const headers = new Headers();
		headers.set("Sec-Fetch-Site", "same-site");
		headers.set("Referer", assignmentUrl);
		await this.fetch(assignmentUrl, { headers });

		return await this.fetch(assignmentUrl, { parseResponse: (text) => text });
	}
}
